### DISCUSSION TITLE
Symbol (or image) changes when selected

### DISCUSSION THREAD

**Discussion by aetkie** [UNC-Project-Open-AAC/OS-DPI#254] [Help]
Looking for ideas on how to make a new design work and my OS-DPI brain is a little rusty.  

IDEA - It's a 3x3 grid. When an icon is selected, I want the image to change. Sot the first image is a star, and when the icon is selected, it's an image of an arrow.

Best to do this in the spreadsheet? Have a column for #symbol and another column for #newsymbol?

Or, write this as an action...?
Origin - grid 
Condition - ?
State - ?
New Value - ?

**Comment by gbishop** [comment]
What happens next? Let's make it easier to think about with letters. The
initial grid is like
a b c
d e f
g h i

You touch the "a" button and the grid becomes
A b c
d e f
g h i

Then what? If I now press "b" what do you want?

**Reply by aetkie** [reply]
if I press b is want 
A B c
d e f
g h i

the end result is 
A B C
D E F 
G H I

**Comment by gbishop** [comment]
There are 512 possible configurations (2 possibilities for each of 9
things, 2 to the 9th power, 2*9 = 512) assuming I can press the buttons in
any order.

I can see how to do it with 9 one-element grids arranged in the 3x3
pattern. Each grid shows either the first or second image depending on the
value of a variable using a filter and there is an action that changes the
value of the variable when the button is pressed. This would require
18 rows in the spreadsheet (2 for each grid) which feels minimal.

It would also require 9 actions and 9 variables.

That would get ugly for a much larger grid but doesn't seem too bad for 3
by 3.

There might be a clever way to use a field with a single variable and a
single action but I can't immediately see it with the filter operations we
have. If we added a "Logical AND" operator to filters, then you could think
in binary and need only 1 field (simply the power of 2 corresponding to the
position in the grid), a single variable and a single action.

With that addition to the filters it might be possible to only have 1
grid...

Does that help?

**Comment by aetkie** [comment]
Yes! That gives me enough to go forward and try more of this on my own. I think I'll try the first solution you suggested, and if I get stuck again, I will post what I have so far. 
Thank you @gbishop

**Comment by aetkie** [comment]
I'm not sure that I have the conditions and updates correct. The button is not changing like I want it to.

[TM_moana_v1.zip](https://github.com/user-attachments/files/16912287/TM_moana_v1.zip)

**Reply by gbishop** [reply]
Here is a 2x2 version of what I intended.
[TM_moana_v2.zip](https://github.com/user-attachments/files/16912990/TM_moana_v2.zip)

**Reply by aetkie** [reply]
Ah, so your solution uses 2 grids. One grid has the question mark (first symbol) and the second grid has the image (new symbol). I like this solution. 
If I added on to your solution, I would just add more grids following the same pattern
13
23...

Why is the state variable $t12? I understand the '$' and the '12' but there does the 't' come from?

**Comment by gbishop** [comment]
No. There are 4 grids. You'll need 9 for 3x3. Each grid controls a
different variable which is also used to filter what the grid displays.

**Comment by gbishop** [comment]
Variable names have to start with a letter so I used t for type

**Reply by aetkie** [reply]
Design complete. The intent for this design was to simulate a simple cause and effect game for an eye gaze user. 
[TM_moana_v2.zip](https://github.com/user-attachments/files/16930752/TM_moana_v2.zip)

**Comment by gbishop** [comment]
I have attached a version that is logically much simpler. I added a new `filter` to Grid to allow testing if a field is contained in a variable.
[TM_moana_v2a.zip](https://github.com/user-attachments/files/17033639/TM_moana_v2a.zip)

Also, none of the sounds are included in the v2 file that I made this from.

**Reply by aetkie** [reply]
I am looking at this version and the sheet. I see the new filter added with the operator "is_contained_in". Where does the state variable $shown come from? Why doesn't it appear in the sheet?

**Comment by gbishop** [comment]
It is a state variable created and updated in the actions
